add support for multipart range requests according to rfc 7233

DanRiess opened the pull request:
This implementation of the WebDav standard was missing multipart range request support. These kinds of requests are important for speeding up NextCloud file access of large binary files.

DanRiess committed:
add support for multipart range requests according to rfc 7233

DanRiess committed:
add support for multipart range requests according to rfc 7233

phil-davis commented:
https://github.com/sabre-io/dav/actions/runs/3134827626/jobs/5090168461
php-cs-fixer complains. Please run:
```
composer install
composer cs-fixer
```
and commit the code-style changes that it makes. That will get past the first bit of CI.

phil-davis commented:
Then try:
```
composer phpunit
```
and sort out getting the unit tests to pass, and add some unit tests to cover new functionality.

DanRiess commented:
Excuse the noob question, but how do I set up the local server for the unit tests? I tried using php -S localhost:8000 in the tests folder (elevated and non elevated console) and failed a lot of them that shouldn't have anything to do with what i wrote. Usually it was some access forbidden or resource not available errors.

phil-davis commented:
After having done `composer install` or `composer update` there will be `sabre/http` in the `vendor` dir. That has some resources that the tests expect to be available at `localhost:8000`

In a terminal from the root of the `dav` repo, run:
```
php -S localhost:8000 -t vendor/sabre/http/tests/www
```

Then run the tests in another terminal window:
```
phil@phil-Inspiron-5468:~/git/sabre-io/dav$ composer phpunit
> phpunit --configuration tests/phpunit.xml
PHPUnit 9.5.25 #StandWithUkraine

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

.............................................................   61 / 1796 (  3%)
.............................................................  122 / 1796 (  6%)
.............................................................  183 / 1796 ( 10%)
.........SSSS................................................  244 / 1796 ( 13%)
.............................................................  305 / 1796 ( 16%)
.............................................................  366 / 1796 ( 20%)
....SSSSSSSSSSSSSSSS.........................................  427 / 1796 ( 23%)
...........................................................SS  488 / 1796 ( 27%)
SSSSSSSSSSSSSSSS.............................................  549 / 1796 ( 30%)
.............................................................  610 / 1796 ( 33%)
.............................................................  671 / 1796 ( 37%)
.............................................................  732 / 1796 ( 40%)
.............................................................  793 / 1796 ( 44%)
.............................................................  854 / 1796 ( 47%)
..................SSSSSSSSSSSSSSSSSSSSSSSSSS.................  915 / 1796 ( 50%)
.............................................................  976 / 1796 ( 54%)
...............SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS 1037 / 1796 ( 57%)
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS. 1098 / 1796 ( 61%)
............................................................. 1159 / 1796 ( 64%)
............................................................. 1220 / 1796 ( 67%)
............................................................. 1281 / 1796 ( 71%)
............................................................. 1342 / 1796 ( 74%)
............................................................. 1403 / 1796 ( 78%)
............................................................. 1464 / 1796 ( 81%)
............................................................. 1525 / 1796 ( 84%)
............................................................. 1586 / 1796 ( 88%)
............................................................. 1647 / 1796 ( 91%)
....................SSSSSSSSSSSSSSSSSSSSSSSSSSSS............. 1708 / 1796 ( 95%)
............................................................. 1769 / 1796 ( 98%)
...........................                                   1796 / 1796 (100%)

Time: 00:11.504, Memory: 102.00 MB

OK, but incomplete, skipped, or risky tests!
Tests: 1796, Assertions: 3183, Skipped: 198.
```

That gets rid of 3 errors that I had at first.

DanRiess committed:
cs-fix, added some unit tests

DanRiess committed:
cs-fix, added some unit tests

DanRiess commented:
Thanks for your help! I did that and it did indeed fix some issues. I still get loads of errors and failures However, i get even more errors when i try to run the unit tests of the original repo. So I guess it's my work environment. I decided to ignore most of them now and only took care about the ones directly affecting get requests and ranges, fixed them and added some unit tests for multipart stuff.

DanRiess committed:
tried fixing codecov complaints

DanRiess committed:
tried fixing codecov complaints

Madrich commented:
I am interested in that feature too. Plz approve ... Many thanks to Daniel

phil-davis committed:
Typos in comments

Madrich commented:
Hi Phil, are you going to merge soon?

Madrich commented:
Please merge asap .... thanks

DeepDiver1975 commented on lib/DAV/Server.php:
```diff
@@ -612,37 +612,191 @@ public function getHTTPDepth($default = self::DEPTH_INFINITY)
-     * If the first offset is null, the second offset should be used to retrieve the last x bytes of the entity
-     *
-     * @return int[]|null
+     * @return string|null
```
This changes the signature of a public function which  breaks BC ...

... we need to see how to work around this ....

DeepDiver1975 replied:
Suggestion:
- keep `getHTTPRange` unchanged
- mark it as deprecated
- add new method `getHttpRangeString` (better naming accepted)
- base the new implementation in core plugin upon the new method

DanRiess replied:
thanks for your feedback! I have added the changes as per your suggestion.

DeepDiver1975 commented on lib/DAV/Server.php:
```diff
@@ -612,37 +612,191 @@ public function getHTTPDepth($default = self::DEPTH_INFINITY)
+     *
+     * @return array|null
+     */
+    public function preprocessRanges($rangeString, $nodeSize)
```
this method can be kept as private in class CorePlugin.

DanRiess committed:
Merge branch 'master' of github.com:DanRiess/sabreDav

DanRiess committed:
added old getHTTPRange function for BC and marked as deprecated, renamed new function to getHTTPRangeHeaderValue, moved preprocessRanges to CorePlugin
